### ISSUE TITLE
Possible bug in BatchNorm layer causing non-reproducible validation losses with LR=0

### ISSUE THREAD

**Issue by abearman** [keras-team/keras#10216]
We discovered that adding BatchNorm layers in a network produces unreproducible validation losses. 

### The set-up
* We used an off-the-shelf ResNet50 network trained on ImageNet, with the last layer chopped off.
* We begin to train this network on the Keras CIFAR10 dataset, using **only 2 images from each of the training and validation set**.
* We use learning rate = 0, and verified that the network weights do not change.
* We only use random horizontal flips for data augmentation (which is only used on the training set).

We would expect that, with two images in the training set and each having a 50% of a horizontal flip, we should see 4 different training losses alternated between randomly (which we **do** observe). The problem is that validation loss fluctuates randomly. I would expect that it should stay constant regardless of data augmentation, because augmentation is not applied to validation images. 

### Other experiments:
We DO NOT observe this erroneous behavior in a network with no BatchNorm layers (including the same ResNet50 network with batch norm layers removed). In this case, training loss fluctuates between 4 constant values and validation loss stays constant every epoch. This is what leads us to believe that the BatchNorm layers are the culprit.

We DO observe this erroneous behavior (albeit with smaller fluctuations in the validation loss) even if no data augmentation is used. In this case, training loss now is the same every epoch, and validation loss fluctuates randomly.

Link to gist with code here: https://gist.github.com/abearman/916673e9a0f12bcf5ac2723b8b5eb819

Run results with data augmentation:
```
Epoch 1/15
1/1 [==============================] - 8s 8s/step - loss: 0.3417 - acc: 0.9000 - val_loss: 0.4247 - val_acc: 0.9000
Epoch 2/15
1/1 [==============================] - 2s 2s/step - loss: 0.3285 - acc: 0.9000 - val_loss: 0.4036 - val_acc: 0.9000
Epoch 3/15
1/1 [==============================] - 2s 2s/step - loss: 0.3043 - acc: 0.9000 - val_loss: 0.4024 - val_acc: 0.9000
Epoch 4/15
1/1 [==============================] - 2s 2s/step - loss: 0.3285 - acc: 0.9000 - val_loss: 0.4049 - val_acc: 0.9000
Epoch 5/15
1/1 [==============================] - 2s 2s/step - loss: 0.3043 - acc: 0.9000 - val_loss: 0.4053 - val_acc: 0.9000
Epoch 6/15
1/1 [==============================] - 2s 2s/step - loss: 0.3417 - acc: 0.9000 - val_loss: 0.4022 - val_acc: 0.9000
Epoch 7/15
1/1 [==============================] - 2s 2s/step - loss: 0.3679 - acc: 0.9000 - val_loss: 0.4021 - val_acc: 0.9000
Epoch 8/15
1/1 [==============================] - 2s 2s/step - loss: 0.3043 - acc: 0.9000 - val_loss: 0.4022 - val_acc: 0.9000
Epoch 9/15
1/1 [==============================] - 2s 2s/step - loss: 0.3285 - acc: 0.9000 - val_loss: 0.4048 - val_acc: 0.9000
Epoch 10/15
1/1 [==============================] - 2s 2s/step - loss: 0.3285 - acc: 0.9000 - val_loss: 0.4053 - val_acc: 0.9000
Epoch 11/15
1/1 [==============================] - 2s 2s/step - loss: 0.3679 - acc: 0.9000 - val_loss: 0.4053 - val_acc: 0.9000
Epoch 12/15
1/1 [==============================] - 2s 2s/step - loss: 0.3285 - acc: 0.9000 - val_loss: 0.4052 - val_acc: 0.9000
Epoch 13/15
1/1 [==============================] - 2s 2s/step - loss: 0.3679 - acc: 0.9000 - val_loss: 0.4056 - val_acc: 0.9000
Epoch 14/15
1/1 [==============================] - 2s 2s/step - loss: 0.3043 - acc: 0.9000 - val_loss: 0.4056 - val_acc: 0.9000
Epoch 15/15
1/1 [==============================] - 2s 2s/step - loss: 0.3679 - acc: 0.9000 - val_loss: 0.4055 - val_acc: 0.9000
```

Run results without data augmentation: (note: we observed larger fluctuations in validation loss for a more complex network)
```
Epoch 1/15
2/2 [==============================] - 8s 4s/step - loss: 0.3296 - acc: 0.9000 - val_loss: 0.3875 - val_acc: 0.9000
Epoch 2/15
2/2 [==============================] - 2s 801ms/step - loss: 0.3296 - acc: 0.9000 - val_loss: 0.3871 - val_acc: 0.9000
Epoch 3/15
2/2 [==============================] - 2s 812ms/step - loss: 0.3296 - acc: 0.9000 - val_loss: 0.3871 - val_acc: 0.9000
Epoch 4/15
2/2 [==============================] - 2s 800ms/step - loss: 0.3296 - acc: 0.9000 - val_loss: 0.3871 - val_acc: 0.9000
Epoch 5/15
2/2 [==============================] - 1s 746ms/step - loss: 0.3296 - acc: 0.9000 - val_loss: 0.3871 - val_acc: 0.9000
Epoch 6/15
2/2 [==============================] - 1s 746ms/step - loss: 0.3296 - acc: 0.9000 - val_loss: 0.3871 - val_acc: 0.9000
Epoch 7/15
2/2 [==============================] - 2s 836ms/step - loss: 0.3296 - acc: 0.9000 - val_loss: 0.3871 - val_acc: 0.9000
Epoch 8/15
2/2 [==============================] - 2s 807ms/step - loss: 0.3296 - acc: 0.9000 - val_loss: 0.3871 - val_acc: 0.9000
```

**Comment by vkk800** [comment]
I might be wrong here, but I think BatchNormalization layer uses moving (population) averages to compute the mean and variance during inference time. They get updated in each epoch when you train the model (even if the learning rate is set to zero) and this is why the output also changes. During training time, BatchNormalization uses mean and variance computed just from the batch which results in the four different output values you observe.

**Comment by isaacgerg** [comment]
My understanding is the mean and variance should not be computed from the batch but learned from the training data.  that is what the exponential moving average is for!
